subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/vcs/VcsDependencyResolver.java

```diff
@@ -71,13 +71,19 @@ public void resolve(DependencyMetadata dependency, BuildableComponentIdResolveRe
         VcsMappingInternal vcsMappingInternal = getVcsMapping(dependency);
 
         if (vcsMappingInternal != null) {
+            // Safe to cast because if it weren't a ModuleComponentSelector, vcsMappingInternal would be null.
+            ModuleComponentSelector depSelector = (ModuleComponentSelector) dependency.getSelector();
             vcsMappingsInternal.getVcsMappingRule().execute(vcsMappingInternal);
 
             // TODO: Need failure handling, e.g., cannot clone repository
             if (vcsMappingInternal.hasRepository()) {
                 VersionControlSpec spec = vcsMappingInternal.getRepository();
                 VersionControlSystem versionControlSystem = versionControlSystemFactory.create(spec);
-                VersionRef selectedVersion = selectVersionFromRepository(spec, versionControlSystem);
+                VersionRef selectedVersion = selectVersionFromRepository(spec, versionControlSystem, depSelector.getVersion());
+                if (selectedVersion == null) {
+                    result.failed(new ModuleVersionResolveException(depSelector, spec.getDisplayName() + " does not contain a version matching " + depSelector.getVersion()));
```

Comment:
This should contain some context about what we were trying to resolve.  It currently looks like:
> Git Repository at file:/Users/sterling/gits/gradle/build/tmp/test%20files/GitVcsIntegrationTest/unknown-test/2tn5u/dep/ does not contain a version matching 1.4.0

It should look like the message from other `ModuleVersionResolveException` failures:
> "Could not resolve $depSelector. Git Repository at file:/Users/sterling/gits/gradle/build/tmp/test%20files/GitVcsIntegrationTest/unknown-test/2tn5u/dep/ does not contain a version matching '1.4.0'

Reply:
Done